components/script/dom/notification.rs

```diff
@@ -0,0 +1,348 @@
+                .ok()
+                .map(|url| USVString::from(url.to_string()))
+        });
+        let badge = options.badge.as_ref().and_then(|badge_url| {
+            ServoUrl::parse_with_base(base_url.as_ref(), badge_url)
+                .ok()
+                .map(|url| USVString::from(url.to_string()))
+        });
+
+        // TODO: vibrate not implemented yet
+        // let vibrate = options.vibrate;
+        let timestamp = options.timestamp.unwrap_or(fallback_timestamp);
+        let renotify = options.renotify;
+        let silent = options.silent;
+        let require_interaction = options.requireInteraction;
+
+        let mut actions: Vec<NotificationAction> = Vec::new();
+        let max_actions = Notification::MaxActions(global);
+        for (idx, action) in options.actions.iter().enumerate() {
+            if idx >= max_actions as usize {
+                break;
+            }
+            actions.push(NotificationAction {
+                action: action.action.clone(), // Spec is using `name`, however in WebIDL it use `action`.
+                title: action.title.clone(),
+                icon: action.icon.clone(),
+            });
+        }
+
+        Ok(Self {
+            eventtarget: EventTarget::new_inherited(),
+            title,
+            body,
+            data,
+            dir,
+            image,
+            icon,
+            badge,
+            lang,
+            silent,
+            origin,
+            timestamp,
+            renotify,
+            tag,
+            require_interaction,
+            actions,
+            closed: Cell::new(false),
+        })
+    }
+}
+
+impl NotificationMethods<crate::DomTypeHolder> for Notification {
+    /// <https://notifications.spec.whatwg.org/#constructors>
+    fn Constructor(
+        global: &GlobalScope,
+        proto: Option<HandleObject>,
+        can_gc: CanGc,
+        title: DOMString,
+        options: RootedTraceableBox<NotificationOptions>,
+    ) -> Fallible<DomRoot<Notification>> {
+        // step 1: Check global is a ServiceWorkerGlobalScope
+        if global.downcast::<ServiceWorkerGlobalScope>().is_some() {
+            return Err(Error::Type("Can not call in a service worker.".to_string()));
+        }
+
+        // step 2: Check options.actions must be empty
+        if !options.actions.is_empty() {
+            return Err(Error::Type("Actions must be empty.".to_string()));
+        }
+
+        // step 3: Create a notification with a settings object
+        // https://notifications.spec.whatwg.org/#create-a-notification-with-a-settings-object
+        let fallback_timestamp = global.performance().Now().to_u64().unwrap();
+        let notification = Notification::new(
+            global,
+            proto,
+            can_gc,
+            title,
+            options,
+            None,
+            None,
+            fallback_timestamp,
+        )?;
+
+        // step 5.1: Check permission
```

Comment:
Please put a quote or a paraphrase from the specification here instead of simply "Check permission" which is pretty ambiguous.